File: mcp-server/excel/ExcelController.java
// ExcelController.java
import com.jacob.activeX.ActiveXComponent;
import com.jacob.com.Dispatch;
import com.jacob.com.Variant;

public class ExcelController {
    private ActiveXComponent excel;

    public ExcelController() {
        excel = new ActiveXComponent("Excel.Application");
        excel.setProperty("Visible", true);
    }

    // ======================
    // 파일
    // ======================
    public void openFile(String path) {
        Dispatch.call(excel.getProperty("Workbooks").toDispatch(), "Open", path);
    }

    public void saveFile() {
        Dispatch.call(excel.getProperty("ActiveWorkbook").toDispatch(), "Save");
    }

    public void closeFile() {
        Dispatch.call(excel.getProperty("ActiveWorkbook").toDispatch(), "Close", false);
    }

    // ======================
    // 셀 / 범위
    // ======================
    public void writeCell(String sheetName, int row, int col, String value) {
        Dispatch sheet = Dispatch.call(excel.getProperty("Sheets").toDispatch(), "Item", sheetName).toDispatch();
        Dispatch.put(Dispatch.call(sheet, "Cells", row, col).toDispatch(), "Value", value);
    }

    public String readCell(String sheetName, int row, int col) {
        Dispatch sheet = Dispatch.call(excel.getProperty("Sheets").toDispatch(), "Item", sheetName).toDispatch();
        return Dispatch.get(Dispatch.call(sheet, "Cells", row, col).toDispatch(), "Value").toString();
    }

    public void mergeCells(String sheetName, int startRow, int startCol, int endRow, int endCol) {
        Dispatch sheet = Dispatch.call(excel.getProperty("Sheets").toDispatch(), "Item", sheetName).toDispatch();
        Dispatch range = Dispatch.call(sheet, "Range",
            String.format("%s%d:%s%d", colToLetter(startCol), startRow, colToLetter(endCol), endRow)).toDispatch();
        Dispatch.put(range, "MergeCells", true);
    }

    // ======================
    // 행 / 열
    // ======================
    public void insertRow(String sheetName, int row) {
        Dispatch sheet = Dispatch.call(excel.getProperty("Sheets").toDispatch(), "Item", sheetName).toDispatch();
        Dispatch rows = Dispatch.get(sheet, "Rows").toDispatch();
        Dispatch rowRange = Dispatch.call(rows, "Item", row).toDispatch();
        Dispatch.call(rowRange, "Insert");
    }

    public void deleteRow(String sheetName, int row) {
        Dispatch sheet = Dispatch.call(excel.getProperty("Sheets").toDispatch(), "Item", sheetName).toDispatch();
        Dispatch rows = Dispatch.get(sheet, "Rows").toDispatch();
        Dispatch rowRange = Dispatch.call(rows, "Item", row).toDispatch();
        Dispatch.call(rowRange, "Delete");
    }

    // ======================
    // 수식
    // ======================
    public void setFormula(String sheetName, int row, int col, String formula) {
        Dispatch sheet = Dispatch.call(excel.getProperty("Sheets").toDispatch(), "Item", sheetName).toDispatch();
        Dispatch cell = Dispatch.call(sheet, "Cells", row, col).toDispatch();
        Dispatch.put(cell, "Formula", formula);
    }

    // ======================
    // 시트
    // ======================
    public void addSheet(String sheetName) {
        Dispatch sheets = excel.getProperty("Sheets").toDispatch();
        Dispatch newSheet = Dispatch.call(sheets, "Add").toDispatch();
        Dispatch.put(newSheet, "Name", sheetName);
    }

    public void deleteSheet(String sheetName) {
        Dispatch sheet = Dispatch.call(excel.getProperty("Sheets").toDispatch(), "Item", sheetName).toDispatch();
        Dispatch.call(sheet, "Delete");
    }

    // ======================
    // 매크로
    // ======================
    public void runMacro(String macroName) {
        Dispatch.call(excel, "Run", macroName);
    }

    // ======================
    // 헬퍼
    // ======================
    private String colToLetter(int col) {
        StringBuilder sb = new StringBuilder();
        while (col > 0) {
            int rem = (col - 1) % 26;
            sb.insert(0, (char)(rem + 'A'));
            col = (col - 1) / 26;
        }
        return sb.toString();
    }
}

File: mcp-server/excel/ExcelController.ts
import { spawn } from "child_process";

export class ExcelController {
  private jarPath: string;

  constructor(jarPath: string = "ExcelController.jar") {
    this.jarPath = jarPath;
  }

  private runMCP(args: string[]): Promise<string> {
    return new Promise((resolve, reject) => {
      const proc = spawn("java", ["-jar", this.jarPath, ...args]);

      let output = "";
      proc.stdout.on("data", (data) => (output += data.toString()));
      proc.stderr.on("data", (data) => console.error(`[Excel MCP ERROR] ${data}`));
      proc.on("close", (code) => {
        if (code === 0) resolve(output.trim());
        else reject(new Error(`Excel MCP exited with code ${code}`));
      });
    });
  }

  // ----- 시트 관리 -----
  async createSheet(sheetName: string) {
    return this.runMCP(["createSheet", sheetName]);
  }
  async deleteSheet(sheetName: string) {
    return this.runMCP(["deleteSheet", sheetName]);
  }
  async renameSheet(oldName: string, newName: string) {
    return this.runMCP(["renameSheet", oldName, newName]);
  }

  // ----- 셀/범위 -----
  async writeCell(sheet: string, cell: string, value: string) {
    return this.runMCP(["writeCell", sheet, cell, value]);
  }
  async readCell(sheet: string, cell: string) {
    return this.runMCP(["readCell", sheet, cell]);
  }
  async clearRange(sheet: string, range: string) {
    return this.runMCP(["clearRange", sheet, range]);
  }

  // ----- 행/열 -----
  async insertRow(sheet: string, rowIndex: number) {
    return this.runMCP(["insertRow", sheet, rowIndex.toString()]);
  }
  async deleteRow(sheet: string, rowIndex: number) {
    return this.runMCP(["deleteRow", sheet, rowIndex.toString()]);
  }
  async insertColumn(sheet: string, colIndex: number) {
    return this.runMCP(["insertColumn", sheet, colIndex.toString()]);
  }
  async deleteColumn(sheet: string, colIndex: number) {
    return this.runMCP(["deleteColumn", sheet, colIndex.toString()]);
  }

  // ----- 서식 -----
  async setCellStyle(sheet: string, cell: string, styleJson: string) {
    return this.runMCP(["setCellStyle", sheet, cell, styleJson]);
  }

  // ----- 수식 -----
  async setFormula(sheet: string, cell: string, formula: string) {
    return this.runMCP(["setFormula", sheet, cell, formula]);
  }

  // ----- 차트 -----
  async createChart(sheet: string, range: string, chartType: string) {
    return this.runMCP(["createChart", sheet, range, chartType]);
  }

  // ----- 테이블 -----
  async createTable(sheet: string, range: string, tableName: string) {
    return this.runMCP(["createTable", sheet, range, tableName]);
  }

  // ----- 매크로 -----
  async runMacro(macroName: string) {
    return this.runMCP(["runMacro", macroName]);
  }
}
